### ISSUE TITLE
False positive `bad-string-format-type` for `bool`, `IntEnum` and other subclasses of the builtin types

### ISSUE THREAD

**Issue by ekanshul** [PyCQA/pylint#11315]
### Bug description

`arg_matches_format_type` in `pylint/checkers/strings.py` compares `arg_type.pytype()` exactly with `builtins.int` / `builtins.float` / `builtins.str`. Instances of subclasses therefore never match, so `bool`, `IntEnum` members and user subclasses of `float`/`int` are reported as not matching `%d`, `%i`, `%x`, `%f`, ... even though they format fine at runtime.

The stdlib hits this in `curses/has_key.py`:

```python
L.append('Mismatch for key %s, system=%i, Python=%i'
         % (_curses.keyname(key), system, python))   # system and python are bools
```

Minimal example:

```python
import enum


class Color(enum.IntEnum):
    RED = 1


class MyFloat(float):
    pass


print("%i" % True)
print("%d %x" % (Color.RED, Color.RED))
print("%f %d" % (MyFloat(1.5), MyFloat(1.5)))
```

### Command used

```shell
pylint --disable=all --enable=bad-string-format-type a.py
```

### Pylint output

```shell
a.py:12:6: E1307: Argument 'builtins.bool' does not match format type 'i' (bad-string-format-type)
a.py:13:6: E1307: Argument 'a.Color.RED' does not match format type 'd' (bad-string-format-type)
a.py:13:6: E1307: Argument 'a.Color.RED' does not match format type 'x' (bad-string-format-type)
a.py:14:6: E1307: Argument 'a.MyFloat' does not match format type 'f' (bad-string-format-type)
a.py:14:6: E1307: Argument 'a.MyFloat' does not match format type 'd' (bad-string-format-type)
```

### Expected behavior

No message: subclasses format exactly like their builtin base type (`bool` and `IntEnum` are `int`s, `MyFloat` is a `float`).

This is the remaining part of #11147 (the `%i`/`%u`/`%a` part was fixed in #11150).

### Pylint version

```shell
pylint 4.1.0-dev0
astroid 4.4.0-dev0
Python 3.13.11
```

### OS / Environment

macOS
